Cluster Profile Tutorial (#1151)
spectrocloud/librarium: content/docs/18-knowledgebase/01-tutorials/10-dev-engine/03-deploy-k8s-cluster.md

Author: Done

The [Spectro Cloud Terraform](https://registry.terraform.io/providers/spectrocloud/spectrocloud/latest/docs) provider enables you to create and manage Palette resources in a codified manner by leveraging Infrastructure as Code (IaC). There are many reasons why you would want to utilize IaC. A few reasons worth highlighting are: the ability to automate infrastructure, improve collaboration related to infrastructure changes, self-document infrastructure through codification, and track all infrastructure in a single source of truth. 

If you need to become more familiar with Terraform, check out the [Terraform explanation from HashiCorp](https://developer.hashicorp.com/terraform/intro). 

<br />

## Prerequisites

To complete this tutorial, you will need the following items

- A Spectro Cloud account.

- Basic knowledge of containers.

- Terraform v1.3.6 or greater.

- A Spectro Cloud API key. [To learn how to create an API key](https://docs.spectrocloud.com/user-management/user-authentication/#apikey)

<br />

> ## Create Provider Account

Author: done

With Terraform, you will create the cluster and deploy the application. Each cluster will be hosted on a cloud service provider, i.e. AWS, Azure, GCP, and managed through Palette.

<br />

## Configure the Cluster Profile

Create a folder where you will put all the Terraform configuration files:
```bash
$ mkdir terraform-profile
$ cd terraform-profile
```

### Providers

Create the file *provider.tf* and insert the content:
```terraform
terraform {
  required_providers {
    spectrocloud = {
      version = "= 0.11.1"
      source  = "spectrocloud/spectrocloud"
    }
  }
}
provider "spectrocloud" {
  project_name = "Default"
  api_key      = var.spectrocloud_api_key
}
```

This file will connect to Palette through the Palette's *api_key*.

<br />

### Variables

Create the file *variables.tf* and insert the content:

```terraform
variable "spectrocloud_api_key" {}
```

Create the file *terraform.tfvars* and insert the content of the variable:

```terraform
spectrocloud_api_key = "j54xxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
```

> To find the value of the *spectrocloud_api_key*, open the Palette dashboard and, from the left **Main Menu** click on the **Tenant Settings** panel and select **API Keys**.

Author: added an env variable

<br />

### Data
The data file contains the details, such as name, version, of the stacks that compose the infrastructure.

Create the file data.tf and insert:

<Tabs>
<Tabs.TabPane tab="AWS" key="aws-tf-profile">

```terraform
data "spectrocloud_pack" "csi" {
  name    = "csi-aws"
  version = "1.0.0"
}
data "spectrocloud_pack" "cni" {
  name    = "cni-calico"
  version = "3.24.5"
}
data "spectrocloud_pack" "k8s" {
  name    = "kubernetes"

>   version = "1.21.14"

Author: For AWS I'm forced to leave the K8s version to "1.21.14" otherwise is not compatible with Amazon Cloud Native Storage (csi-aws)

    values = data.spectrocloud_pack.cni.values
  }
  pack {
    name   = data.spectrocloud_pack.csi.name
    tag    = data.spectrocloud_pack.csi.version
    uid    = data.spectrocloud_pack.csi.id
    values = data.spectrocloud_pack.csi.values
  }
  pack {
    name   = data.spectrocloud_pack.proxy.name
    tag    = data.spectrocloud_pack.proxy.version
    uid    = data.spectrocloud_pack.proxy.id
    values = data.spectrocloud_pack.proxy.values
  }
}
```

## Create the Profile

To create the cluster profile on Palette use the Terraform commands to apply the information present in the configuration files.

Open the terminal and enter into the folder where you have the Terraform configuration files

```bash
$ cd terraform-profile
```

Initialize the working directory having Terraform configuration files
```bash
$ terraform init
```

Validate the configuration files in the directory
```bash
$ terraform validate 
```

Create the execution plan with the preview changes your configuration will create on the infrastructure
```bash
$ terraform plan
```

Finally, apply the modifications there are in the plan to execute them and create the infrastructure
```bash
$ terraform apply
```
<br />

> 

Author: Done

## Verify the Profile

<Tabs>

> <Tabs.TabPane tab="AWS" key="aws-validation-p">

Author: Done

To check the profile creation on Palette, login to Palette dashboard and, from the left **Main Menu** click on the **Profiles** panel to access the profile page. At the top of the list you can find the *tf-azure-profile*:

![Terraform Azure profile](/tutorials/deploy-clusters/terraform/tf_azure_profile.png)

<br />

Click on the profile to see the details of the stacks that compose the profile:

![Terraform Azure profile details](/tutorials/deploy-clusters/terraform/tf_azure_profile_details.png)

</Tabs.TabPane>
<Tabs.TabPane tab="Google Cloud" key="gcp-validation-p">

To check the profile creation on Palette, login to Palette dashboard and, from the left **Main Menu** click on the **Profiles** panel to access the profile page. At the top of the list you can find the *tf-gcp-profile*:

![Terraform GCP profile](/tutorials/deploy-clusters/terraform/tf_gcp_profile.png)

<br />

Click on the profile to see the details of the stacks that compose the profile:

![Terraform GCP profile details](/tutorials/deploy-clusters/terraform/tf_gcp_profile_details.png)

</Tabs.TabPane>
</Tabs>

<br />

## Configure the Cluster

<Tabs>
<Tabs.TabPane tab="AWS" key="aws-tf-cluster">

> Create a folder where you will put all the Terraform configuration files:

Author: Done

```bash
$ mkdir terraform-cluster
$ cd terraform-cluster
```

### Variables

Create the file *variables.tf* and insert the list of variables:

```terraform

> variable "spectrocloud_api_key" {}

Author: done

variable "cluster_profile" {}
variable "region" {}
variable "aws_ssh_key_name" {}
variable "aws-cloud-account-name" {
    type = string
    description = "The name of your AWS account as assigned in Palette"
}
variable "master_nodes" {
    type = object({
        count           = string
        instance_type   = string
        disk_size_gb    = string
        availability_zones = list(string)
    })
    description = "Master nodes configuration."
}
variable "worker_nodes" {
    type = object({
        count           = string
        instance_type   = string
        disk_size_gb    = string
        availability_zones = list(string)
    })
    description = "Worker nodes configuration."
}
```

Then, create also the file *terraform.tfvars* and append the content of the variables:

```terraform
spectrocloud_api_key    = "j54xxxxxxxxxxxxxxxxxxxxxxxxxxxx"

> aws_ssh_key_name        = "aws-key"

Author: Done

aws-cloud-account-name  = "82xxxxxxxxxx"
cluster_profile         = "tf-profile"
region                  = "us-east-1"
master_nodes = {
    count           = "1"
    instance_type   = "m5.large"
    disk_size_gb    = "60"
    availability_zones = ["us-east-1a"]
}
worker_nodes = {
    count           = "2"
    instance_type   = "m5.large"
    disk_size_gb    = "60"
    availability_zones = ["us-east-1a"]
}
```
<br />

### Cluster Resources

Create the *cluster.tf* file and insert the cluster resources definition:

```terraform
data "spectrocloud_cloudaccount_aws" "account" {
  name = var.aws-cloud-account-name
}
data "spectrocloud_cluster_profile" "profile" {
  name = var.cluster_profile

> }

Author: All the Terraform files moved to one folder and modified the config files accordingly.

data "spectrocloud_cloudaccount_gcp" "account" {
  name = var.gcp-cloud-account-name
}
data "spectrocloud_cluster_profile" "profile" {
  name = var.cluster_profile
}
resource "spectrocloud_cluster_gcp" "cluster" {
  name             = "gcp-cluster"
  tags             = ["gcp", "tutorial"]
  cloud_account_id = data.spectrocloud_cloudaccount_gcp.account.id
  cloud_config {
    project = "Default"
    region  = var.region
  }
  cluster_profile {
    id = data.spectrocloud_cluster_profile.profile.id
  }
  machine_pool {
    control_plane           = true
    control_plane_as_worker = true
    name                    = "master-pool"
    count                   = var.master_nodes.count
    instance_type           = var.master_nodes.instance_type
    disk_size_gb            = var.master_nodes.disk_size_gb
    azs                     = var.master_nodes.availability_zones
  }
  machine_pool {
    name          = "worker-basic"
    count         = var.worker_nodes.count
    instance_type = var.worker_nodes.instance_type
    disk_size_gb  = var.worker_nodes.disk_size_gb
    azs           = var.worker_nodes.availability_zones
  }
}
```

</Tabs.TabPane>
</Tabs>

<br />

## Create the Cluster

> 

Author: Done

To create the cluster on the cloud provider use the Terraform commands to apply the information present in the configuration files.

Open the terminal and enter into the folder where you have the Terraform configuration files

```bash
$ cd terraform-cluster
```

Initialize the working directory having Terraform configuration files
```bash
$ terraform init
```

Validate the configuration files in the directory
```bash
$ terraform validate 
```

Create the execution plan with the preview changes your configuration will create on the infrastructure
```bash
$ terraform plan
```

Finally, apply the modifications there are in the plan to execute them and create the infrastructure
```bash
$ terraform apply
```
<br />

## Verify the Cluster

<Tabs>
<Tabs.TabPane tab="AWS" key="aws-validation">

> To check the cluster creation, login to Palette dashboard and, from the left **Main Menu** click on the **Clusters** panel from the left panel and check the created cluster. At the top of the list you can find the *aws-cluster*:

Author: done

        ports:
        - containerPort: 8080
```

In this code example, we deploy the [*hello-universe*](https://github.com/spectrocloud/hello-universe) demo application.

We set two replicas for the application to simulate a distributed environment with a redundant web application deployed on Kubernetes. In front of them, we add a load balancer service to route requests across all replica containers as best practice to maximize the workload and to expose a single access point to the web application.

For more information about the service LoadBalancer component you can refer to the [Kubernetes official documentation](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer).

<br />

Then, use Terraform to push the modification to Palette and finalize the application deployment.

So, first, check the validation of the configuration files:
```bash
$ terraform validate
```

Create the execution plan with the additional modifications to the already existant configuration:
```bash
$ terraform plan
```

Finally, apply the modifications there are in the plan to execute them and create the infrastructure:
```bash
$ terraform apply
```

> 

Reviewer: This can all be reduced to a single command `terraform apply --auto-approve`

Author: done

</Tabs.TabPane>
</Tabs>

<br />

> ### Deploy

Reviewer: This step is not needed with Terraform. If you keep everything in the same folder, the cluster will receive the new update. This problem will go away when you simplify everything under two main tabs UI workflow and Terraform.

Author: Removed for Terraform

![Deployed application](/tutorials/deploy-clusters/app.png)

<br />

# Clean-up

It is a good practice to clean up the resources used for the tutorial from the cloud provider, to avoid unexpected cloud charges.

<br />

<Tabs>
<Tabs.TabPane tab="UI Workflow" key="ui-clean">

## UI Workflow

To remove the cluster created in this tutorial, open the Palette dashboard and, from the left **Main Menu** click on the **Cluster** panel to access the clusters page. Select the cluster you want to delete to access its details page.

Click on **Settings**, at the top-right corner of the page, from the details page to expand the settings menu and select **Delete Cluster** to delete the cluster.

![Destroy-cluster](/tutorials/deploy-clusters/delete-cluster-button.png)

You will be asked to type in the cluster name to confirm the delete action. Continue and type the cluster name to proceed with the delete step. Repeat this process for all the cluster you want to delete.

</Tabs.TabPane>
<Tabs.TabPane tab="Terraform" key="terraform-clean">

## Terraform

> To destroy the cluster with Terraform, use the destroy command.

Reviewer: ```suggestion
Use the destroy command to remove all the resources you created through Terraform.
```

First enter into the folder where you have the Terraform configuration for the cluster
```bash
$ cd terraform-cluster

> ```

Author: done

If you want to check the resources you will delete, you can first execute: 
```terraform
terraform plan -destroy
```

Then delete the components by running the destroy command:
```terraform

> $ terraform destroy

Reviewer: ```suggestion
terraform destroy --auto-approve
```

```

> Wait until it finishes to delete the cluster.

Reviewer: ```suggestion
The deletion process will take a few minutes. 
```
Let's also include the output that shows the count of resources removed and let's add a note about force delete. 

![CleanShot 2023-04-07 at 10 09 48](https://user-images.githubusercontent.com/29551334/230649185-1dca3de1-7670-40cc-b2b9-6c69665fae14.png)
